// decktet-utils/src/decktet-utils.js
var _ = require('lodash')

const Suits = [ "Warrior", "Mage", "Rogue", "Cleric" ],
      Ranks = [ "1", "2", "3", "4", "5", "6"],
      Types = [ "C","S"],
      HeroCards = [
    		{ rank: "1", suits: [ "Warrior" ], Type: "C", Name: "Warrior Level 1", Text:"" },
    		{ rank: "1", suits: [ "Mage" ], Type: "C", Name: "Mage Level 1", Text:"" },
    		{ rank: "1", suits: [ "Rogue" ], Type: "C", Name: "Rogue Level 1", Text:""},
    		{ rank: "1", suits: [ "Cleric" ], Type: "C", Name: "Cleric Level 1", Text:""},

    		{ rank: "2", suits: [ "Warrior" ], Type: "C", Name: "Warrior Level 2", Text:""},
    		{ rank: "2", suits: [ "Mage" ], Type: "C" , Name: "Mage Level 2", Text:""},
    		{ rank: "2", suits: [ "Rogue" ], Type: "C" , Name: "Rogue Level 2", Text:""},
    		{ rank: "2", suits: [ "Cleric" ], Type: "C" , Name: "Cleric Level 2", Text:""},

    		{ rank: "3", suits: [ "Warrior" ], Type: "C" , Name: "Warrior Level 3", Text:""},
    		{ rank: "3", suits: [ "Mage" ], Type: "C" , Name: "Mage Level 3", Text:""},
    		{ rank: "3", suits: [ "Rogue" ], Type: "C" , Name: "Rogue Level 3", Text:""},
    		{ rank: "3", suits: [ "Cleric" ], Type: "C", Name: "Cleric Level 3", Text:"" },

    		{ rank: "4", suits: [ "Warrior" ], Type: "C" , Name: "Warrior Level 4", Text:""},
    		{ rank: "4", suits: [ "Mage" ], Type: "C", Name: "Mage Level 4", Text:"" },
    		{ rank: "4", suits: [ "Rogue" ], Type: "C" , Name: "Rogue Level 4", Text:""},
    		{ rank: "4", suits: [ "Cleric" ], Type: "C", Name: "Cleric Level 4", Text:"" },

    		{ rank: "5", suits: [ "Warrior" ], Type: "C", Name: "Warrior Level 5", Text:"" },
    		{ rank: "5", suits: [ "Mage" ], Type: "C" , Name: "Mage Level 5", Text:""},
    		{ rank: "5", suits: [ "Rogue" ], Type: "C" , Name: "Rogue Level 5", Text:""},
    		{ rank: "5", suits: [ "Cleric" ], Type: "C" , Name: "Cleric Level 5", Text:""},
    		
    		{ rank: "6", suits: [ "Warrior" ], Type: "C", Name: "Warrior Level 6", Text:"" },
    		{ rank: "6", suits: [ "Mage" ], Type: "C" , Name: "Mage Level 6", Text:""},
    		{ rank: "6", suits: [ "Rogue" ], Type: "C" , Name: "Rogue Level 6", Text:""},
    		{ rank: "6", suits: [ "Cleric" ], Type: "C" , Name: "Cleric Level 6", Text:""},
    		
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Knowledge is Power", Text:"Draw two cards at the end of the turn."  },
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Silence I want to read", Text:"Your oponent can't play cards this turn. They draw an adittional card."  },
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Everything is better on Fire", Text:"Your cards have level +1. Play a card. If you can't play a card, draw two cards." },
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Rainbow Union", Text:"Your cards have any color and you play two cards. The resulting level is the sum."  }
    	],
    	
      MonsterCards = [
    		{ rank: "1", suits: [ "Warrior" ], Type: "C", Name: "Cat", Text:"" },
    		{ rank: "1", suits: [ "Mage" ], Type: "C", Name: "Dark Cat", Text:"" },
    		{ rank: "1", suits: [ "Rogue" ], Type: "C", Name: "Flying Cat", Text:""},
    		{ rank: "1", suits: [ "Cleric" ], Type: "C", Name: "Cat Zombie", Text:""},

    		{ rank: "2", suits: [ "Warrior" ], Type: "C", Name: "Kobold", Text:""},
    		{ rank: "2", suits: [ "Mage" ], Type: "C" , Name: "Frozen Kobold", Text:""},
    		{ rank: "2", suits: [ "Rogue" ], Type: "C" , Name: "Kobold Trap", Text:""},
    		{ rank: "2", suits: [ "Cleric" ], Type: "C" , Name: "Kobold Ghost", Text:""},

    		{ rank: "3", suits: [ "Warrior" ], Type: "C" , Name: "Goblin", Text:""},
    		{ rank: "3", suits: [ "Mage" ], Type: "C" , Name: "Icy Goblin", Text:""},
    		{ rank: "3", suits: [ "Rogue" ], Type: "C" , Name: "Flying Goblin", Text:""},
    		{ rank: "3", suits: [ "Cleric" ], Type: "C", Name: "Kobold Zombie", Text:"" },

    		{ rank: "4", suits: [ "Warrior" ], Type: "C" , Name: "Goblin Knight", Text:""},
    		{ rank: "4", suits: [ "Mage" ], Type: "C", Name: "Mimic", Text:"" },
    		{ rank: "4", suits: [ "Rogue" ], Type: "C" , Name: "Trap Treasure", Text:""},
    		{ rank: "4", suits: [ "Cleric" ], Type: "C", Name: "Goblin Skeleton", Text:"" },

    		{ rank: "5", suits: [ "Warrior" ], Type: "C", Name: "Orc", Text:"" },
    		{ rank: "5", suits: [ "Mage" ], Type: "C" , Name: "Magic Orc", Text:""},
    		{ rank: "5", suits: [ "Rogue" ], Type: "C" , Name: "Balloon Orc", Text:""},
    		{ rank: "5", suits: [ "Cleric" ], Type: "C" , Name: "Undead Orc", Text:""},
    		
    		{ rank: "6", suits: [ "Warrior" ], Type: "C", Name: "Giant Bear", Text:"" },
    		{ rank: "6", suits: [ "Mage" ], Type: "C" , Name: "Giant Bear with Glasses", Text:""},
    		{ rank: "6", suits: [ "Rogue" ], Type: "C" , Name: "Big Bat", Text:""},
    		{ rank: "6", suits: [ "Cleric" ], Type: "C" , Name: "Vampire Cat Zombie", Text:""},
    		
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Knowledge is Power", Text:"Draw two cards at the end of the turn."  },
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Silence I want to read", Text:"Your oponent can't play cards this turn. They draw an adittional card."  },
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Everything is better on Fire", Text:"Your cards have level +1. Play a card. If you can't play a card, draw two cards." },
    		{ rank: "0", suits: [ "Warrior","Mage","Rogue","Cleric" ], Type: "S", Name: "Rainbow Union", Text:"Your cards have any color and you play two cards. The resulting level is the sum."  }
    	]

function countBySuit(cards) {
	var counts = {}
	cards.forEach(function(card) {
		card.suits.forEach(function(suit) {
			counts[suit] = ( counts[suit] || 0 ) + 1
		})
	})
	return counts
}

function countByRank(cards) {
	var counts = {};
	cards.forEach(function(card) {
		counts[card.rank] = ( counts[card.rank] || 0 ) + 1;
	});
	return counts;
}

function shareOneSuit(cards) {
	// se algum suit estiver presente na mesma quantidade de cartas que
	// a selecao, entao todas possuem aquele suit
	counts = countBySuit(cards);
	for ( suit in counts )
		if ( counts[suit] == cards.length )
			return suit;
	return null;
}

function shareOneRank(cards) {
	counts = countByRank(cards);
	for ( rank in counts )
		if ( counts[rank] == cards.length )
			return rank;
	return null;
}

function hasOneOfEachSuit(cards) {
	var counts = countBySuit(cards);
	// counts deve ser um objeto com 1 em cada um dos seis suits
	return allSuits.every(function(suit) {
		return counts[suit] == 1;
	});
}

function areConsecutiveRanks(cards, comparator) {
	// Cria um novo array e ordena pelo "rank traduzido"
	var clone = cards.filter(function(e) { return true; });
	clone.sort(rankComparator(Ranks));

	// Verifica se toda carta vem depois da sua antecessora
	if ( comparator == null )
		var comparator = rankComparator(Ranks);

	for ( var i = 1; i < clone.length; i++ ) {
		var difference = comparator(clone[i-1], clone[i] )
		if ( difference != -1 )
			return false; // uma carta nao eh sucessora da outra
	}
	return true;
}

function findCard(fullName) {

	var pieces = fullName.split("of");
	var rank = _.camelCase(pieces[0]);

	pieces = pieces[1].replace(",", " and ").split("and");
	var suits = pieces.map(function(el) { return _.camelCase(el) });

	matches = Cards.filter(function(card) { return card.rank == rank && _.intersection(card.suits, suits).length == suits.length });
	return matches.length ? matches[0] : null;
}

function makeDeck(fullNames) {
	return fullNames.map(findCard);
}

// http://stackoverflow.com/questions/6274339/how-can-i-shuffle-an-array-in-javascript
function shuffle(cards) {
  for(var j, x, i = cards.length; i; j = Math.floor(Math.random() * i), x = cards[--i], cards[i] = cards[j], cards[j] = x);
  return cards;
}

function rankComparator(arr) {
	return function(cardA, cardB) {
		if ( _.isString(cardA) ) cardA = findCard(cardA);
		if ( _.isString(cardB) ) cardB = findCard(cardB);
		return arr.indexOf(cardA.rank) - arr.indexOf(cardB.rank);
	}
}

function cardName(card) {
	return card.Name
}

module.exports = {
  Ranks, Suits, Types, MonsterCards, HeroCards,
  countBySuit, countByRank,
  shareOneSuit, shareOneRank,
  rankComparator, cardName, shuffle,
  makeDeck, findCard, areConsecutiveRanks,
  hasOneOfEachSuit
}
